### ISSUE TITLE
How can I differ the credit card gateway from offsite one?

### ISSUE THREAD

**Issue by makasim** [thephpleague/omnipay#309]
The credit card gateways requires different behavior than the offsite one. So if I want to change one gateway with another one, the client has to be able to differ them and act differently.

**Comment by makasim** [comment]
@kayladnls ping

**Comment by judgej** [comment]
This is true. You generally need to write code that is specific to each gateway and the way it behaves. This is because every gateway has some it is own surprises and special ways of doing things that it brings. What is not possible at the moment, is to write a generic wrapper for which you can drop in any gateway so that it *just works*.

OmniPay deals mainly with sending and receiving messages, and abstracts the components (fields, meaning of statues) as much as possible. This helps to reduce the code you need to write, but you still need to understand how the gateway works in order to put the right process/flow/logic in place.

**Comment by judgej** [comment]
Do you have any specific gateways you had in mind?

**Comment by makasim** [comment]
stripe and paypal for example

**Comment by makasim** [comment]
> What is not possible at the moment, is to write a generic wrapper for which you can drop in any gateway so that it just works.

I am trying to achieve it here https://github.com/Payum/OmnipayBridge. Almost done.

**Comment by makasim** [comment]
This question is related to this PR https://github.com/Payum/OmnipayBridge/pull/25

**Comment by kayladnls** [comment]
@judgej is right, however I think your suggestion of `supportsCompletePurchase` is in the right vein, as only offsite gateways would need to use such a thing.

**Comment by delatbabel** [comment]
Unfortunately not *all* offsite gateways have a completePurchase() method.

Furthermore some gateways are both offsite and credit card based. PayPal_Rest is an example -- it can be used with or without a credit card number (for USA and UK based merchants).  First Data have a gateway that is similar.

So I don't think that there is a method to determine, in advance of making a purchase, whether a gateway is redirect or not. The only sure fire way is to send a purchase() call and then test isRedirect() on the response.  Or store a flag in your own code, for example.

**Comment by makasim** [comment]
I pretty much abstracted it in the bridge. 

> The only sure fire way is to send a purchase() call and then test isRedirect() on the response. Or store a flag in your own code, for example.

@delatbabel It is too late for me, since if the gateway needs a credit card, it already threw an exception. Asking user for a credit all the time even if it is a offsite gateway looks odd to me too. So I have three gateways: universal, credit card, and offsite, First one tries to guess what omnipay gateway it wraps where the credit card and offsite ones forced to use those methods respectively. 

One more note. If it is  a credit card gateway but user does not provide it, I have an abiliy to ask user to provide us with it.

**Comment by delatbabel** [comment]
> So I have three gateways: universal, credit card, and offsite

That sounds like what you need.

What I do with my own code is to store a gateway object in the database with some configuration data attached. That configuration data includes a JSON blob of the initialisation parameters for the gateway (2 blobs actually, one for test mode and one for production), some data about what currencies the gateway supports, and some flags about whether the gateway supports direct (credit card) payments or redirect payments or other sort (e.g. manual).

There really isn't anything you can do otherwise. None of the gateways support a method which will return any kind of response saying what sort of gateway it is.
